clients/src/main/java/org/apache/kafka/common/requests/OffsetsForLeaderEpochResponse.java

```diff
@@ -51,133 +41,82 @@
+
+    public OffsetsForLeaderEpochResponse(OffsetForLeaderEpochResponseData data) {
+        this.data = data;
     }
 
-    private final int throttleTimeMs;
-    private final Map<TopicPartition, EpochEndOffset> epochEndOffsetsByPartition;
-
-    public OffsetsForLeaderEpochResponse(Struct struct) {
-        this.throttleTimeMs = struct.getOrElse(THROTTLE_TIME_MS, DEFAULT_THROTTLE_TIME);
-        this.epochEndOffsetsByPartition = new HashMap<>();
-        for (Object topicAndEpocsObj : struct.get(TOPICS)) {
-            Struct topicAndEpochs = (Struct) topicAndEpocsObj;
-            String topic = topicAndEpochs.get(TOPIC_NAME);
-            for (Object partitionAndEpochObj : topicAndEpochs.get(PARTITIONS)) {
-                Struct partitionAndEpoch = (Struct) partitionAndEpochObj;
-                Errors error = Errors.forCode(partitionAndEpoch.get(ERROR_CODE));
-                int partitionId = partitionAndEpoch.get(PARTITION_ID);
-                TopicPartition tp = new TopicPartition(topic, partitionId);
-                int leaderEpoch = partitionAndEpoch.getOrElse(LEADER_EPOCH, RecordBatch.NO_PARTITION_LEADER_EPOCH);
-                long endOffset = partitionAndEpoch.get(END_OFFSET);
-                epochEndOffsetsByPartition.put(tp, new EpochEndOffset(error, leaderEpoch, endOffset));
-            }
-        }
+    public OffsetsForLeaderEpochResponse(Struct struct, short version) {
+        data = new OffsetForLeaderEpochResponseData(struct, version);
+    }
+
+    public OffsetsForLeaderEpochResponse(Map<TopicPartition, EpochEndOffset> offsets) {
+        this(0, offsets);
     }
 
-    public OffsetsForLeaderEpochResponse(Map<TopicPartition, EpochEndOffset> epochsByTopic) {
-        this(DEFAULT_THROTTLE_TIME, epochsByTopic);
+    public OffsetsForLeaderEpochResponse(int throttleTimeMs, Map<TopicPartition, EpochEndOffset> offsets) {
+        data = new OffsetForLeaderEpochResponseData();
+        data.setThrottleTimeMs(throttleTimeMs);
+
+        offsets.forEach((tp, offset) -> {
+            OffsetForLeaderTopicResult topic = data.topics().find(tp.topic());
+            if (topic == null) {
+                topic = new OffsetForLeaderTopicResult().setTopic(tp.topic());
+                data.topics().add(topic);
+            }
+            topic.partitions().add(new OffsetForLeaderPartitionResult()
+                .setPartition(tp.partition())
+                .setErrorCode(offset.error().code())
+                .setLeaderEpoch(offset.leaderEpoch())
+                .setEndOffset(offset.endOffset()));
+        });
     }
 
-    public OffsetsForLeaderEpochResponse(int throttleTimeMs, Map<TopicPartition, EpochEndOffset> epochsByTopic) {
-        this.throttleTimeMs = throttleTimeMs;
-        this.epochEndOffsetsByPartition = epochsByTopic;
+    public OffsetForLeaderEpochResponseData data() {
+        return data;
     }
 
     public Map<TopicPartition, EpochEndOffset> responses() {
```

Comment:
I hope we can get rid of those conversion in the future :)

Reply:
I will address this in a follow-up PR.